### PR TITLE
Refactor the 'go' function to accept options

### PR DESCRIPTION
**Pull request by andreogle**
### Description

Currently we have a `go` function that returns a tuple of `[Error | null, T | null]` when given a function. This function was pretty basic and needed to be wrapped when we wanted to retry or timeout.

After a suggestion from Emanuel, I thought we should adjust this function to accept an optional "options" object where things like `retries` can be specified. The interface looks like:

```ts
export interface PromiseOptions {
  retries?: number;
  retryDelayMs?: number;
  timeoutMs?: number;
}
```

So now you can call `go()` with a much nicer interface:

```ts
const myPromise = () => Promise.resolve('it worked');
const [err, res] = await go(myPromise, { retries: 1 });
```

### CODE REVIEW

**Comment by sitch** [comment]
Looks good -- if it's helpful there are a some libs out there that do similar things, use at your own discretion:

https://github.com/franckLdx/ts-retry
https://www.npmjs.com/package/@lifeomic/attempt
https://www.npmjs.com/package/fetch-retry
https://github.com/normartin/ts-retry-promise

File: packages/node/src/utils/promise-utils.ts [hunk 1]
```diff
@@ -87,6 +68,28 @@ export interface ContinuousRetryOptions {
...
+// duration has been reached which is a problem.
+export function promiseTimeout<T>(ms: number, promise: Promise<T>): Promise<T> {
+  return Bluebird.resolve(promise).timeout(ms);
+}
+
+export function wait(ms: number) {
+  return new Promise((r) => setTimeout(r, ms));
+}
+
 export function retryOnTimeout<T>(maxTimeoutMs: number, operation: () => Promise<T>, options?: ContinuousRetryOptions) {
```
Comment (amarthadan) [inline packages/node/src/utils/promise-utils.ts:93]: Is this function still needed?

**Review by amarthadan — CHANGES_REQUESTED** [review]
I was about to agree with @sitch on this one and suggest using some library that's already doing this but I can't find one that's combining go-like return format and retries :sweat_smile: But still, maybe it would be worth to rework this as a separate library since we're using it for the ChainAPI as well and the implementations will inevitably differ.

File: packages/node/src/utils/promise-utils.ts (hunk 1 above)
Comment (andreogle) [reply packages/node/src/utils/promise-utils.ts:93]: Yeah, but it's only used when executing an API call at the moment

File: packages/node/src/utils/promise-utils.ts [hunk 2]
```diff
@@ -9,71 +9,52 @@ export function isPromise(obj: any) {
 
 type Response<T> = [Error, null] | [null, T];
 
+export interface PromiseOptions {
+  retries?: number;
+  retryDelay?: number;
+  timeoutMs?: number;
```
[the remark this answers is not in the document]
Comment (andreogle) [reply packages/node/src/utils/promise-utils.ts:15]: Good catch. If there is no `timeoutMs` given, then I don't think it shouldn't apply a timeout

File: packages/node/src/utils/promise-utils.ts [hunk 3]
```diff
@@ -9,71 +9,52 @@ export function isPromise(obj: any) {
 
 type Response<T> = [Error, null] | [null, T];
```
[the remark this answers is not in the document]
Comment (andreogle) [reply packages/node/src/utils/promise-utils.ts]: Renamed to `GoResult` but open to better names

**Comment by andreogle** [comment]
In general I try to avoid adding random smaller dependencies if we can implement the functionality ourselves easily, especially when they aren't widely used or updated semi-reguarly

https://www.npmjs.com/package/@lifeomic/attempt looks like it could be a decent package to include though from the list above.

It probably makes sense to extract these utils into a separate package but that can be in a separate PR. Logged an issue here: https://github.com/api3dao/airnode/issues/347

**Comment by andreogle** [comment]
I replaced the custom `retryOperation` implementation with https://www.npmjs.com/package/@lifeomic/attempt It seems to work well, is better tested and offers more utilities if we ever need them

**Comment by sitch** [comment]
More of a side-note but Id recommend a general caution around overly convenient retry syntax particularly if it makes things so easy that implicit retries are turned on by default. 

For HTTP, network level retries are probably more likely to create intermittent/prod-only issues than stabilize the system. If somehow logging/eventing gets partially bypassed (like a successful retry being logged the same as a single successful request), you can create some non-obvious / more involved problem solving scenarios. 

The other thing is that having retries so close to the metal can make call prioritization impossible or with bad heuristics (i.e. blocking retries dominating valid calls). 

I generally recommend just re-enqueuing the failed request back into a (priority) queue/event log. Makes for a much simpler mental model with a greater ability to manage the micro/macro/situational.

**Comment by andreogle** [comment]
@sitch thanks, but not sure I fully understand everything you've said here.

> implicit retries are turned on by default.

What are "implicit retries"? Nothing is retried by default either, it's always opt-in.

> like a successful retry being logged the same as a single successful request

This is how it works at the moment too, but it's not ideal. If a HTTP fails first then succeeds on the second, you would probably want to know about both. It's tricky because we need to log additional context too (`coordinatorId`, `chain` etc). I think we might need a custom implementation to return both the value and errors that happened

> I generally recommend just re-enqueuing the failed request back into a (priority) queue/event log.

How would this work in the context of Airnode? i.e. no readily available database/cache and things moving through a sort of step-by-step pipeline?

**Comment by sitch** [comment]
> What are "implicit retries"? Nothing is retried by default either, it's always opt-in.

Ah good! I was just saying that with nice syntax this can sometimes just get 1 line of code in somewhere that can easily go unnoticed.  That said, the one line,  in most cases leads to a system most users would probably deem "more stable".  The hidden problem is that it can add complexity to our mental model as engineers -- a kind of Win/Lose.  Better to increase stability AND simplify the mental model -- Win/Win :)

> How would this work in the context of Airnode? i.e. no readily available database/cache and things moving through a sort of step-by-step pipeline?

It's all just data structures at the end of the day.  Databases are really just hyper-graphs and don't really do all that much for systems that seek to be highly available & distributed.  

One way clean / high throughput system would be:

1. Accept request
1. Create a sum type with a call context as: `{request, context}` (ensure this is serializable)
1. Enqueue it into a priority queue/heap (really anything that produces a rank/order/group/cluster over it)
1. Use pull-based demand to pop off the highest priority items (batch size proportional to connection pool).
1. Re-enqueue fails with modified call contexts, otherwise mark entry as unsuccessful 
1. Spin up workers based on queue wait times, queue size, or any metric you want

**Comment by sitch** [comment]
@andreogle you can see how in ~20 lines of elixir you can achieve the above: https://github.com/dashbitco/broadway

**Comment by andreogle** [comment]
Ok thanks. Sounds like that would be a pretty significant rework of the way things are currently done though and definitely way outside the scope of this PR 😉 

I doubt we'll see many problems with the current implementation of how retrying is implemented. Most of these retries are simply for when a call to the blockchain provider errors. e.g. the call to get API call templates fails because of some network reason.

**Pull request merged**